src/app/clients/add/databaseFunctions.ts

```diff
@@ -46,6 +55,203 @@ export const insertClient = async (
+const deleteFailedInsert = async (primaryKey: string): Promise<void> => {
+    await supabase.from("clients").delete().eq("primary_key", primaryKey);
+};
+
+const updateClient = async (
+    clientRecord: ClientDatabaseUpdateRecord,
+    primaryKey: string
+): Promise<ClientDatabaseFetchRecord> => {
+    const {
+        data: ids,
+        status,
+        error,
+    } = await supabase
+        .from("clients")
+        .update(clientRecord)
+        .eq("primary_key", primaryKey)
+        .select("primary_key, family_id");
+
+    if (error === null && Math.floor(status / 100) === 2) {
+        return ids![0];
+    }
+    throw Error(
+        `Error occurred whilst updating into Clients table. HTTP Code: ${status}, PostgreSQL Code: ${error?.code}. `
+    );
+};
+
+const updateChild = async (child: Person): Promise<void> => {
+    const childRecord: FamilyDatabaseUpdateRecord = {
+        gender: child.gender,
+        age: child.age,
+    };
+    const { status, error } = await supabase
+        .from("families")
+        .update(childRecord)
+        .eq("primary_key", child.primaryKey);
+
+    if (error !== null || Math.floor(status / 100) !== 2) {
+        throw Error(
+            `Error occurred whilst updating the Families table. HTTP Code: ${status}, PostgreSQL Code: ${error?.code}. `
+        );
+    }
+};
+
+const countAdultMember = async (familyID: string, gender: string): Promise<number> => {
+    const { count, error } = await supabase
+        .from("families")
+        .select("*", { count: "exact", head: true })
+        .eq("family_id", familyID)
+        .eq("gender", gender)
+        .is("age", null);
+
+    if (error !== null || count === null) {
+        throw error;
+    }
+    return count;
+};
+
+const deleteAdultMembers = async (
+    familyID: string,
+    gender: string,
+    count: number
+): Promise<void> => {
+    const { status, error } = await supabase
+        .from("families")
+        .delete()
+        .eq("family_id", familyID)
+        .eq("gender", gender)
+        .is("age", null)
+        .limit(count);
+
+    if (error !== null || Math.floor(status / 100) !== 2) {
+        throw Error(
+            `Error occurred whilst updating the Families table. HTTP Code: ${status}, PostgreSQL Code: ${error?.code}. `
+        );
+    }
+};
+
+const updateFamily = async (
+    adults: Person[],
+    children: Person[],
+    familyID: string
+): Promise<void> => {
+    for (const adultGroup of adults) {
+        const numberAdultsInDatabase = await countAdultMember(familyID, adultGroup.gender);
+        const difference = adultGroup.quantity! - numberAdultsInDatabase;
+        if (difference > 0) {
+            const arrayToInsert: Person[] = [{ ...adultGroup, quantity: difference }];
+            await insertFamily(arrayToInsert, familyID);
+        }
+        if (difference < 0) {
+            await deleteAdultMembers(familyID, adultGroup.gender, difference);
+        }
+    }
+
+    children.forEach(updateChild);
```

Comment:
Not too sure if this and the updateChild function is doing what it is intended to do. What if the number of children change?